<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
    <xacro:include filename="colors.xacro" />
    <!-- xacro:include filename="inertial_macros.xacro" / -->

    
    <xacro:property name="base_mass"        value="0.5"  />
    <xacro:property name="base_length"      value="0.6"  />
    <xacro:property name="base_width"       value="0.3"  />
    <xacro:property name="base_height"      value="0.1"  />

    <xacro:property name="wheel_mass"       value="0.1"  />
    <xacro:property name="wheel_radius"     value="0.05" />
    <xacro:property name="wheel_width"      value="0.05"  />
    <xacro:property name="clearance"        value="0.05" />    
    <xacro:property name="wheel_separation" value="0.4"  />
    <xacro:property name="wheel_dist_x"     value="0.4"  />


    
    <xacro:macro name="box_link" params="name size color color_rgb mass mass_x mass_y mass_z" >
        <link name="${name}">
            <visual>
                <geometry>
                    <box size="${size}"/>
                </geometry>
                <material name="${color}">
                    <color rgba="${color_rgb}"/>
                </material>
            </visual>

            <collision>
                <geometry>
                    <box size="${size}"/>
                </geometry>
            </collision>

            <xacro:inertial_box mass="${mass}" x="${mass_x}" y="${mass_y}" z="${mass_z}">
                <origin xyz="${mass_x} ${mass_y} ${mass_z}" rpy="0 0 0"/>
            </xacro:inertial_box>                
        </link>
    </xacro:macro>name

    <xacro:macro name="wheel" params="name xyz">
        <link name="${name}_link">
            <visual>
                <geometry>
                    <cylinder radius="${wheel_radius}" length="${wheel_width}" />
                </geometry>

                <material name="red"/>
            </visual>

            <collision>
                <geometry>
                    <cylinder length="${wheel_width}" radius="${wheel_radius}" />
                </geometry>

                <material name="blue"/>
            </collision>

            <xacro:inertial_cylinder mass="${wheel_mass}" length="${wheel_width}" radius="${wheel_radius}">
                <origin xyz="0 0 0" rpy="0 0 0"/>
            </xacro:inertial_cylinder>              
        </link>

        <joint name="${name}_joint" type="continuous">
            <parent link="base_link" />
            <child link="${name}_link" />
            <origin xyz="${xyz}" rpy="${pi/2} 0 0"/>
            <axis rpy="0 0 0" xyz="0 0 1"/>
        </joint>

        <gazebo reference="${name}_link">
            <material>Gazebo/Red</material>
        </gazebo>
    </xacro:macro>  

    <!-- xacro:macro name="static_joint" params="name parent child">
        <joint name="${name}" type="continuous">
            <parent link="${parent}" />
            <child link="${child}" />
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </joint>
    </xacro:macro -->

    <!-- <xacro:macro name="wheel_joint" params="name parent child xyz ">
        <joint name="${name}" type="continuous">
            <parent link="${parent}" />
            <child link="${child}" />
            <origin xyz="${xyz}" rpy="${pi/2} 0 0"/>
            <axis rpy="0 0 0" xyz="0 0 1"/>
        </joint>
    </xacro:macro>    -->

    <!-- Robot Description -->

    <link name="base_link">
    </link>         

    <!-- <xacro:box_link name="chassis" size="0.6 0.3 0.1" color="green" color_rgb="0 1 0 1"  -->
    <xacro:box_link name="chassis" size="${base_length} ${base_width} ${base_height}" color="green" color_rgb="0 1 0 1" 
         mass="0.5" mass_x="0.0" mass_y="0.0" mass_z="0.0"/>

    <joint name="chassis_joint" type="fixed">
        <parent link="base_link"/>
        <child link="chassis"/>
        <origin xyz="0 0 0"/>
    </joint>


    <xacro:wheel name="wheel_fr" xyz="${ wheel_dist_x/2} ${ wheel_separation/2} ${-(clearance - wheel_radius + base_height/2)}"/>
    <xacro:wheel name="wheel_fl" xyz="${ wheel_dist_x/2} ${-wheel_separation/2} ${-(clearance - wheel_radius + base_height/2)}"/>
    <xacro:wheel name="wheel_br" xyz="${-wheel_dist_x/2} ${ wheel_separation/2} ${-(clearance - wheel_radius + base_height/2)}"/>
    <xacro:wheel name="wheel_bl" xyz="${-wheel_dist_x/2} ${-wheel_separation/2} ${-(clearance - wheel_radius + base_height/2)}"/>


    <!-- <xacro:wheel name="right_wheel_frontside" mass="0.1" />
    <xacro:wheel_joint name="base_link_right_wheel_frontside" parent="base_link" 
        child="right_wheel_frontside" xyz="0.2 -0.2 -0.05" />

    <xacro:wheel name="right_wheel_backside" mass="0.1" />
    <xacro:wheel_joint name="base_link_right_wheel_backside" parent="base_link" 
        child="right_wheel_backside" xyz="-0.2 -0.2 -0.05" />

    <xacro:wheel name="left_wheel_frontside" mass="0.1" /> 
    <xacro:wheel_joint name="base_link_left_wheel_frontside" parent="base_link" 
        child="left_wheel_frontside" xyz="0.2 0.2 -0.05" />

    <xacro:wheel name="left_wheel_backside" mass="0.1" /> 
    <xacro:wheel_joint name="base_link_left_wheel_backside" parent="base_link" 
        child="left_wheel_backside" xyz="-0.2 0.2 -0.05" /> -->

    <!-- Gazebo Colors -->
    <!-- <gazebo reference="chassis">
        <material>Gazebo/Green</material>
    </gazebo>
    <gazebo reference="right_wheel_frontside">
        <material>Gazebo/Red</material>
    </gazebo>
    <gazebo reference="right_wheel_backside">
        <material>Gazebo/Red</material>
    </gazebo>
    <gazebo reference="left_wheel_frontside">
        <material>Gazebo/Red</material>
    </gazebo>
    <gazebo reference="left_wheel_backside">
        <material>Gazebo/Red</material>
    </gazebo> -->

</robot>

